doc/sphinx/proof-engine/vernacular-commands.rst

```diff
@@ -929,6 +903,11 @@ Quitting and debugging
       If the given :n:`@command` succeeds, then :n:`Fail @command`
       fails with this error message.
 
+.. note:: The commands that include a :n:`@command` (currently
+   :cmd:`Time`, :cmd:`Redirect`, :cmd:`Timeout` and :cmd:`Fail`) can't
+   use :cmd:`Drop`, :cmd:`Quit`, :cmd:`BackTo`, :cmd:`Show Goal`
+   or :cmd:`Show Proof` as the :n:`@command`.  This is a syntatic restriction.
+   The documentation describes all of these as :n:`@command`\s for simplicity.
```

Comment:
It would be great to gather all these commands that are only usable in the toplevel in a single section.

Reply:
If you mean :cmd:`Drop`, :cmd:`Quit`, :cmd:`BackTo`, :cmd:`Show Goal` and :cmd:`Show Proof`, I must disagree.  The limitation is an implementation detail (and shouldn't even apply to the two Show commands--that's not at all what I wanted).  They should appear in sections based on what they do.

Reply:
This is not an accident that `Drop`, `Quit`, `BackTo` can only be used in `coqtop`, this is on purpose. These are special commands to control the state of the document.

Contrary to what you wrote, `Show Proof` is compatible with `Fail`, it's just `Show Proof Diffs` which isn't, and you are right that this is an accident that this variant arrived in this category, but nonetheless, it is important to highlight that this is only usable in `coqtop` and thus that IDE users must use a similar function provided by their IDE instead.

Finally, regarding `Show Goal`, it is clearly a command for internal use only since you must know state ID numbers to use it.

Anyway, now that I think about it again, I agree that gathering the latter two commands with the first three is not a good idea. But I also come to think that this note is quite useless as well, since most users won't be able to use these commands anyway.

What wouldn't be useless would be to note that for the query commands that accept a goal selector, or for the tactics, `Fail` *et al.* come *in front* of the goal selector. And they must also come *in front* of attributes. In fact, the commands should really not take a `@command` as an argument but a `@vernacular`.

Reply:
> the commands should really not take a `@command` as an argument but a `@vernacular`.

The selector syntax is not entirely correct.  I'd rather not address that in this PR.  So I will leave it as `@command` for now.

Reply:
What is the issue with replacing the `@command` with `@vernacular`?

Reply:
It would take me time to figure out why I had that impression or if it's OK.  Too many parallel threads in this PR, many of which require research.  I've been at this for at least the last 11 hours.

Reply:
Well, I'm pretty sure this is OK and better than `command`.

> I've been at this for at least the last 11 hours.

Take it easy, there's no rush to finish this chapter. I've tried to provide quicker feedback to you than it used to be the case, but the faster you react, the more work it is for me too :stuck_out_tongue_closed_eyes:

Reply:
I guess we could use `@vernacular`.  I don't fully follow how the code works for `noedit_mode` and `tactic_mode`, so a bit hesitant about that.  Probably not worth going into that now.

I think we should extend `vernacular` to show `toplevel_selector` and perhaps to break out `query_command`, referencing a ```:production:`query_command` ``` in the Query commands section.  Also I still tend to think we should cover this note in the grammar; the word picture describing where the goal selector and attributes go is less understandable than a grammar:

```
For :cmd:`Time`, :cmd:`Redirect`, :cmd:`Timeout` and :cmd:`Fail`,
the goal selector and any attributes that apply to the :n:`@command` must
appear after the outer command name (e.g. after `Time`).
```

We should also accurately describe `Info`, which can take a selector and allows ending the sentence with  `...`.  What does the ellipsis do?  The selector can only be `par:` or, like query_command, a single number?

Of course, we've not done the chapters covering selectors or Ltac.

And FWIW not quite done with the grammar for BULLET and `{` ... `}` in proofs, either.

What do you think?

Reply:
> Take it easy, there's no rush to finish this chapter.

Indeed, let's not overdo it.  It's a marathon, not a sprint.

Reply:
> I guess we could use `@vernacular`. I don't fully follow how the code works for `noedit_mode` and `tactic_mode`, so a bit hesitant about that. Probably not worth going into that now.
> 
> I think we should extend `vernacular` to show `toplevel_selector` and perhaps to break out `query_command`, referencing a `` :production:`query_command` `` in the Query commands section. Also I still tend to think we should cover this note in the grammar; the word picture describing where the goal selector and attributes go is less understandable than a grammar:
> 
> ```
> For :cmd:`Time`, :cmd:`Redirect`, :cmd:`Timeout` and :cmd:`Fail`,
> the goal selector and any attributes that apply to the :n:`@command` must
> appear after the outer command name (e.g. after `Time`).
> ```

Maybe we should rename this strange `vernacular` non-terminal into something that's easier to understand, e.g. `sentence`. Then it's quite easy to explain and to understand that a Coq document is composed of Coq sentences that end with dots. `Fail`, `Redirect`, etc. are sentence transformers that take a complete Coq sentence (including any attribute and goal selector) and return a Coq sentence (whose meaning is possibly altered).

> We should also accurately describe `Info`, which can take a selector and allows ending the sentence with  `...`.  What does the ellipsis do?  The selector can only be `par:` or, like query_command, a single number?

Forget about `Info` for the moment. This "command" is an heresy. It starts with a capital letter, but morally speaking it is just a tactical (i.e. a tactic transformer). From a grammar perspective, it behaves exactly as any tactic. It accepts any kind of goal selector. The trailing `...` are not specific to `Info`. They are accepted in proof mode after any tactic if you start your proof using `Proof with @tactic.` The fact that they are hard-coded in the grammar of `Info` (and accepted even without `Proof with`) is yet another compatibility quirk.

Reply:
I added `sentence` as well as `query_command` and `control_command`.  Think this is both accurate and readily understood.  We can deal with `tactic_mode: | OPT toplevel_selector "{"` and `Info` later.

I covered the ` Proof with` case for `...`.  This seems like a rather limited feature.  Is it important or a bit of a legacy feature/syntactic sugar?

BTW, it looks like CoqIDE doesn't handle `...`.  It get stuck after the first period; I can't get it to process anything after it in the buffer (using 8.10.1):

![image](https://user-images.githubusercontent.com/1253341/78615323-0703c780-7826-11ea-97e3-96e72e83b4df.png)

Reply:
This is parsed correctly for me both in 8.10.2 and 8.11.0. If you manage to reproduce this bug in a more recent version, it would be good to report it.

But no, this is not an important feature. On the contrary, this is a feature that most Coq developers probably disapprove today. I once tried to use it in the stdlib and was discouraged from using this style. It requires special UI support, for something that in the end could be replaced by `Ltac z := auto.` and `;z.` instead of `...` (same number of characters). It would be interesting to run a CI experiment to estimate the impact of removing this feature.

Reply:
Yeah, it works correctly in 8.11.0.

In the `theories` directory I count 47 occurrences of `Proof with`  and 405 occurrences of `...` (a few of which are in comments).  Maybe worthwhile removing those uses at some point--users (and I) may use them as stylistic clues.  We could discuss deprecating/removing the feature, too.  I could submit an issue.

Reply:
>users (and I) may use them as stylistic clues

What do you mean?

Reply:
Meaning I search through the .v files and if I see something that appears useful, I emulate it.  *Many* months ago you mentioned that those files use some older, no longer recommended tactics.  Would be better if the .v files used best practices.

Reply:
OK, got it. Indeed, you are right. The only thing we ensure in the stdlib is that it does not use deprecated features.

Reply:
Is it worth replacing the `...` and/or adding a note for `Proof with` about alternative?  Or adding it to the list of enhancements identified during the doc update (as I suggested above)?  Perhaps such a list would be useful for considering relative priority.

Reply:
It makes sense to put it in the list. It may not indeed have top priority.